### ISSUE TITLE
Incorrect lifetime elision recommendation by clippy

### ISSUE THREAD

**Issue by vmalloc** [rust-lang-nursery/rust-clippy#3284]
Trying to lint the following code:

```rust
fn format_error<'a>(e: &'a Error) -> Cow<'a, str> {
...
}
```
Clippy gives:
```
error: explicit lifetimes given in parameter types where they could be elided
```
However the lifetime here is essential to the definition of the `Cow` return type.

```
$ cargo clippy -V
0.0.212
```

Thanks in advance!

**Comment by vmalloc** [comment]
My bad. The answer is 
```
fn format_error(e: &Error) -> Cow<'_, str>
```

**Comment by oli-obk** [comment]
We should improve the diagnostic here

**Comment by vmalloc** [comment]
@oli-obk thanks! much appreciated. Is this something a relative beginner can craft a PR for or is this deep in clippy's innards?

**Comment by oli-obk** [comment]
Oh yes, definitely.

You should go find the code that emits said error and use the `span_note` method (look for other uses of it to see how to call it) to emit an additional message about either removing the lifetimes entirely or using `'_` where a lifetime is required

**Comment by vmalloc** [comment]
@oli-obk sounds like a plan. Will do!

**Comment by vmalloc** [comment]
@oli-obk does it make sense to just fix the note's text like 
```
            "explicit lifetimes given in parameter types where they could be elided (or replaced with '_ )",
```
Or should a more thorough change be made that detects the lifetimes being used and suggest where to place that change? The latter seems like more work than this merits, and might require some refactoring...

**Comment by vmalloc** [comment]
I opened a preliminary PR for this. I hope it's in the right direction
